src/main/scala/com/example/PersistentSagaActor.scala

```diff
@@ -128,9 +133,19 @@ class PersistentSagaActor(persistentEntityRegion: ActorRef)
     case StartSaga(transactionId, commands) =>
       log.info(s"starting new saga with transactionId: $transactionId")
       state = state.copy(currentState = Pending, commands = commands)
+      // FIXME the usage of snapshots in this PersistentActor doesn't look good
+      // - snapshots should only be used as an optimization to avoid replaying too many events
+      // - snapshot for each state change is not efficient, PersistentActor and journals are optimized for event sourcing
+      // - it's not safe, since it's not guaranteed to load the latest snapshot in next recovery
       saveSnapshot(state)
       context.become(pending.orElse(stateReporting))
+      // FIXME use Timers API instead https://doc.akka.io/docs/akka/current/actors.html#timers-scheduled-messages
       context.system.scheduler.scheduleOnce(retryAfter, self, Retry)
+
+      // FIXME I don't think this will scale. Starting one such eventsByTag query for each saga.
+      // Perhaps we should move most of the saga logic into an event processor instead?
+      // We can have a few (fixed number) of those and they process all transactions (creating child actor for
+      // each transactionId). See https://github.com/akka/akka-samples/pull/82
```

Comment:
another thought:  think the problem is that you must limit the number of subscribers (eventsByTag queries). Perhaps having one such per node would be an simple solution. Then it could even publish those events to the local system.eventStream, which the actual saga actors can listen to.